cs/evidence.go

```
@@ -0,0 +1,116 @@
+		OriginalProof json.RawMessage `json:"originalProof"`
+		CurrentProof  json.RawMessage `json:"currentProof"`
+	}{}
+	json.Unmarshal(data, &serialized)
+	deserializer, exists := DeserializeMethods[serialized.Backend]
+	if !exists {
+		return errors.New("Evidence type does not exists")
+
+	}
+	originalProof := deserializer(serialized.OriginalProof)
+	currentProof := deserializer(serialized.CurrentProof)
+	*e = Evidence{
+		State:         serialized.State,
+		Backend:       serialized.Backend,
+		Provider:      serialized.Provider,
+		OriginalProof: originalProof,
+		CurrentProof:  currentProof,
+	}
+	return nil
+}
+
+// Proof is the generic interface which a custom proof type has to implement
+type Proof interface {
+	Time() uint64            // returns the timestamp (UNIX format) contained in the proof
+	FullProof() []byte       // returns data to independently validate the proof
+	Verify(interface{}) bool // Checks the validity of the proof
+}
+
+// GenericProof implements the Proof interface
+type GenericProof struct {
+	Timestamp uint64      `json:"timestamp"`
+	Data      interface{} `json:"data"`
+	Pubkey    []byte      `json:"pubkey"`
+	Signature []byte      `json:"signature"` //sign(hash(time+data))
+}
+
+// Time returns the timestamp from the block header
+func (p *GenericProof) Time() uint64 {
+	return p.Timestamp
+}
+
+// FullProof returns a JSON formatted proof
+func (p *GenericProof) FullProof() []byte {
+	bytes, err := json.MarshalIndent(p, "", "   ")
+	if err != nil {
+		return nil
+	}
+	return bytes
+}
+
+// Verify returns true if the proof of a given linkHash is correct
+func (p *GenericProof) Verify(_ interface{}) bool {
+	return true
+}
+
+// init needs to define a way to deserialize a DummyProof
+func init() {
+	DeserializeMethods["generic"] = func(rawProof json.RawMessage) Proof {
+		p := GenericProof{}
+		if err := json.Unmarshal(rawProof, &p); err != nil {
+			return nil
+		}
+		return &p
+	}
+}
```

Don't you think it might be problematic later that we swallow errors at this level ? Don't we want to return them (I thought that was the Go best practice)

---

good point indeed